A Material Design Makeover for Substrate Front-End Template

stojanov-igor opened the pull request:
#### Project Abstract

This project aims to enhance the user experience of the [substrate-front-end-template](https://github.com/substrate-developer-hub/substrate-front-end-template) by migrating to Material Design UI, ensuring a visually appealing, intuitive, and modern interface that aligns with contemporary design standards.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [x] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ Polkadot (USDC & USDT) address in the application).
- [x] I am aware that, in order to receive a grant, I (and the entity I represent) have to successfully complete a KYC/KYB check.
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

stojanov-igor commented:
I have read and hereby sign the Contributor License Agreement.

keeganquigley reviewed (changes requested):
Thanks for the application @stojanov-igor it would be nice to see this worked on again but I have a few questions:

- Material 3 is pretty cool. Have you looked into [Polkadot Cloud](https://polkadot.cloud/)? Since it is frame-agnostic, I'm wondering if you could integrate the [library](https://github.com/polkadot-cloud/library) to keep the themes/aesthetics in line with other Polkadot apps such as the [staking dashboard](https://staking.polkadot.network/).
- Regarding M1, we usually don't fund the initial designs. Could you remove that deliverable?
- Because of this, we typically ask for mock-ups and/or figma designs to be provided upfront. Could you add these even if they are a rough draft?

stojanov-igor commented:
Hi @keeganquigley 

> Material 3 is pretty cool. Have you looked into [Polkadot Cloud](https://polkadot.cloud/)? Since it is frame-agnostic, I'm wondering if you could integrate the [library](https://github.com/polkadot-cloud/library) to keep the themes/aesthetics in line with other Polkadot apps such as the [staking dashboard](https://staking.polkadot.network/).

I do agree that using the common theme colors is ideal. I will integrate the  polkadot [cloud themes](https://polkadot.cloud/using-themes). 

> Regarding M1, we usually don't fund the initial designs. Could you remove that deliverable?

Removed. 
> Because of this, we typically ask for mock-ups and/or figma designs to be provided upfront. Could you add these even if they are a rough draft?

Added an initial draft to the original proposal application.

muddlebee commented:
It will be nice to see Angular + Material templates for the Substrate Front-End Template and Polkadot Cloud stuff :)

keeganquigley reviewed (dismissed):
Sorry for the delay @stojanov-igor and thanks for making the changes. Given the cost and the fact that we had to close a [previous grant](https://github.com/w3f/Grants-Program/pull/1601) for this, I'm willing to go ahead with it. 

~~While waiting for others to comment, we also [started requiring KYC/KYB checks](https://grants.web3.foundation/docs/faq#what-is-kyckyb-and-why-do-i-have-to-provide-this-data) for all potential grantees. Could you please provide the information outlined under [this link](https://in.sumsub.com/idensic/l/#/uni_mjcB05gJZfh8dbr7)? Let me know if you have any questions or issues!~~

takahser commented on applications/si-front-end-template.md:
```diff
@@ -0,0 +1,148 @@
+| 1. | Cards Module | Implement dashboard cards with basic statistics about the blockchain. |
+| 2. | Table Module| Add table that holds Account balances that can be loaded from Polkadot Extension |
+| 3. | Transfer Module | Simple module to transfer tokens from one acount to another |
+| 4. | Upgrade Runtime Module| EEnsure users can upload wasm File from the UI |
```
```suggestion
| 4. | Upgrade Runtime Module| Ensure users can upload a wasm File from the UI that allows them to upgrade their runtime |
```

takahser replied:
is this a new feature?

takahser commented on applications/si-front-end-template.md:
```diff
@@ -0,0 +1,148 @@
+| **0c.** | Docker | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| 2. | Menu Header | UI element that replaces existing menu items |
+| 3. | Footer | New Element for implementing a new UI Element footer |
+| 4. | Polkadot Extension| Ensure working connection with Polkadot extension |
```
```suggestion
| 4. | Polkadot Extension | Ensure working connection with Polkadot extension |
```

takahser replied:
Is this a new feature?

takahser commented on applications/si-front-end-template.md:
```diff
@@ -0,0 +1,148 @@
+| 2. | Menu Header | UI element that replaces existing menu items |
+| 3. | Footer | New Element for implementing a new UI Element footer |
+| 4. | Polkadot Extension| Ensure working connection with Polkadot extension |
+| 5. | Light/Dark Mode| Enable styling switch to dark and light mode |
```
```suggestion
| 5. | Light/Dark Mode | Enable styling switch to dark and light mode |
```

takahser commented on applications/si-front-end-template.md:
```diff
@@ -0,0 +1,148 @@
+| 2. | Table Module| Add table that holds Account balances that can be loaded from Polkadot Extension |
+| 3. | Transfer Module | Simple module to transfer tokens from one acount to another |
+| 4. | Upgrade Runtime Module| EEnsure users can upload wasm File from the UI |
+| 5. | Pallet Interactor | Module that allows interaction with extrinsics and queries |
+| 6. | Event Module | Displaying latest events that happens on-chain
```
Are these new features?

stojanov-igor replied:
Hi @takahser ,

These are not new features, but UI redisin of existing modules.

takahser reviewed (changes requested):
@stojanov-igor thanks, this looks interesting. I've got a few questions:

- are you going to use any existing frontend library for the implementation?
- are you sure you'd need 2 months full-time to implement this relatively simply UI changes, i.e. replacing all UI elements with their Material-based counterparts? I've taken a brief look at the code and I have the impression that it could be rather quick and easy. Am I missing something here, i.e. is there anything that requires a bit more attention and is more complex than just replacing some UI elements and wiring up their props/state hooks?
- would you be open to rewrite the components to TypeScript as part of this grant?
- are you going to fully overhaul the template and use material for any UI component, or are there any exceptions?
- finally, feel free to have a look at my inline comments. There I'm mainly interested if these features you list bring any new or updated functionality, or if the scope is limited to UI element updates.

stojanov-igor commented:
Hi @keeganquigley 

Thanks for your comments. I have already completed a KYC. @semuelle  requested some changes that I have already incorporated in the proposal. 

@takahser Thanks for the minor improvements. You will find answers to your questions below. 

> are you going to use any existing frontend library for the implementation?

Yes. I will use individual web components from the mui framework. `@mui/material @emotion/react @emotion/styled`

> are you sure you'd need 2 months full-time to implement this relatively simply UI changes, i.e. replacing all UI elements with their Material-based counterparts? I've taken a brief look at the code and I have the impression that it could be rather quick and easy. Am I missing something here, i.e. is there anything that requires a bit more attention and is more complex than just replacing some UI elements and wiring up their props/state hooks?

Based on my analysis there are about 20 distinct components that have to be restyled in the Material Design Style Guide, so the estimated 2-month timeframe is appropriate. Some of the existing components (such as the top dashboard cards) don't have an exact implementation in Material Design, so it would require custom work. As mentioned in the proposal, there is 
 a new functionality that adds themes. 


> would you be open to rewrite the components to TypeScript as part of this grant?

No

> are you going to fully overhaul the template and use material for any UI component, or are there any exceptions?
finally, feel free to have a look at my inline comments. There I'm mainly interested if these features you list bring any new or updated functionality, or if the scope is limited to UI element updates.

It will be a full overhaul of the template based on Material Design Standards. In the case when these standards are not provided, I will provide custom styling that is closest to the provided standard. Most of the features that are listed already exist, so it will be mostly UI updates to modernize the front-end template in hopes of enabling new upcoming developers a better user experience when building and working with Substrate.

takahser reviewed (changes requested):
@stojanov-igor thanks for your reply. In that case I'm going to withhold my vote for now, the reasons being:

- I feel like 2 month is a lot to replace some ui-components + some props/state hook updates with little additional logic
- regarding the top dashboard card: I think the only thing missing is the horizontal line, which should be easy to add: you can just add a `<hr />` HTML element or set the CSS `border` attribute accordingly.
- While I applaud the idea of giving the template a fresh look, I believe a rewrite to TypeScript would be useful as well, since it would introduce better type and code safety, IDE support, readability, etc. In case you change your mind on this, I'd be open to support the grant.
- Themes are nice but I'm not sure if it's very useful, since this template is used for the very early-stages of a Polkadot SDK project, rather than a public/customer-facing product.

keeganquigley reviewed (changes requested):
Thanks for the explanations @stojanov-igor I do agree with my colleague that TypeScript would be better, and this might help to change the minds of other committee members. If not willing to switch, would you be able to reduce the scope/cost a bit?

semuelle reviewed (commented):
Thanks for the updates, @stojanov-igor. I have unresolved two issues from @takahser above, as it seems to me that the questions weren't answered. 
My main question - maybe I missed the answer in your document - would be whether you are planning to have these changes merged into the substrate-front-end-template repo or whether you are planning to maintain a fork. For the former, have you reached out to the maintainers, (e.g. @sacha-l) if this is something they'd be interested in?

takahser commented:
@semuelle regarding the issues you've unresolved above - I think the features mostly exist already, quote from @stojanov-igor's [comment](https://github.com/w3f/Grants-Program/pull/2159#issuecomment-1884989122):

> Most of the features that are listed already exist, so it will be mostly UI updates to modernize the front-end template in hopes of enabling new upcoming developers a better user experience when building and working with Substrate.

stojanov-igor commented:
Hi all,

My comments/ answers are below. 

> would be whether you are planning to have these changes merged into the substrate-front-end-template repo or whether you are planning to maintain a fork. For the former, have you reached out to the maintainers, (e.g. @sacha-l) if this is something they'd be interested in?

The plan is to maintain a fork of the official substrate-front-end repository under the same open and permissive license. We can create a pull request from the fork to the official repository if the maintainers would like to adopt my changes.

> Thanks for the explanations @stojanov-igor I do agree with my colleague that TypeScript would be better, and this might help to change the minds of other committee members.

If the community prefers a typescript version, we will do the conversion to typescript as part of the delivery. Some old school javascript developers dislike strict typing and type annotations but if typescript is a strong preference for future development, we are willing to provide a typescript version. 

>  I feel like 2 month is a lot to replace some ui-components + some props/state hook updates with little additional logic
> regarding the top dashboard card: I think the only thing missing is the horizontal line, which should be easy to add: you can just add a HTML element or set the CSS border attribute accordingly.
> While I applaud the idea of giving the template a fresh look, I believe a rewrite to TypeScript would be useful as well, since it would introduce better type and code safety, IDE support, readability, etc. In case you change your mind on this, I'd be open to support the grant.
> Themes are nice but I'm not sure if it's very useful, since this template is used for the very early-stages of a Polkadot SDK project, rather than a public/customer-facing product.

Thank you for expressing your opinion @takahser. I think you are underestimating the work required. For instance, if you only look at the pallet [Interactor](https://github.com/substrate-developer-hub/substrate-front-end-template/blob/main/src/Interactor.js)  you will notice that these are not simple components. The fact that the code has not been updated in more than 2 years is also worrying for those who want to build applications on polkadot. 

As mentioned above, we can provide a typescript rewrite if this is what the development community wants.

takahser commented:
@stojanov-igor thanks for your reply

> if you only look at the pallet [Interactor](https://github.com/substrate-developer-hub/substrate-front-end-template/blob/main/src/Interactor.js) you will notice that these are not simple components.

Pls correct me if I'm wrong, but wouldn't you just have to update the `return` value with the new UI components? Meaning, you'd bind all the event handlers of the existing component to the new ones (e.g. `onChange={onInterxTypeChange}`) and you're done.

> As mentioned above, we can provide a typescript rewrite if this is what the development community wants.

That would be lovely indeed. I'd support the grant, if a complete TypeScript rewrite was included.

stojanov-igor commented:
Hi all,

I have added Typescript migration as part of the milestone 1. 

> Pls correct me if I'm wrong, but wouldn't you just have to update the return value with the new UI components? Meaning, you'd bind all the event handlers of the existing component to the new ones (e.g. onChange={onInterxTypeChange}) and you're done.

The template is currently using semantic-ui-react. Migrating to Material-Ui is a little more complex than changing the attributes to new ones due to different design philosophies of Semantic UI and Material UI. For instance, the Interactor component that I mentioned previously is using `Grid, Form, Dropdown, Input, Label` which might not have the same attributes within Material Design. Furthermore, there are custom components that the substrate team has implemented such as `TxButton, TxGroupButton` within `substrate-lib` which have to be restyled manually. 

Also, Material-UI follows a tree-shaking approach for importing components which allows developers to selectively import and use only the needed components. Semantic UI has a pre-build bundle which is larger in-size and can negatively affect performance. 

Hope that clears out the confusion.

keeganquigley reviewed (approved):
Thanks for being willing to switch to TS, happy to re-approve.

takahser reviewed (approved):
> Also, Material-UI follows a tree-shaking approach for importing components which allows developers to selectively import and use only the needed components. Semantic UI has a pre-build bundle which is larger in-size and can negatively affect performance.

I understand that there are some design differences, but the library already comes with it, right? Sure, you'd have to adjust some imports, but I don't see how that would significantly increase the workload. But anyway, since you've added the TypeScript rewrite I'm fine with approving this as well.
